New tag added for windows

lprat opened the pull request:
## One line description of pull request
I added new tag for windows (tagging plugin).


## Description:

Just new tag for windows (tagging plugin).

**Related issue (if applicable):** fixes #<plaso issue number here>

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/master/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [ ] Automated checks (Travis, Codecov, Codefactor )pass
  - [ ] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [ ] Reviewer assigned

joachimmetz commented:
@lprat thx for the PR, I'll have a look as soon as the AppVeyor tests are running again  log2timeline/l2tdevtools#794

joachimmetz commented:
log2timeline/l2tdevtools#794 has been resolved and tests are pasing.

@lprat can you write/reference to some accompanying documentation regarding these new tagging rules. E.g. "user_suspect_file" why a ".lnk" file is suspect. This might not be clear to all plaso users and this information might get lost easily over time.

lprat commented:
Sorry for my late reply. Can i add information (to explain) in Analysis-plugin-tagging.md ?
I am trying to do it next week!

joachimmetz commented:
> Sorry for my late reply. Can i add information (to explain) in Analysis-plugin-tagging.md ?

Yes please do, at least let's capture the intent/idea behind the tagging rules

> I am trying to do it next week!

Great, thx, no hurries

lprat commented:
I hope this will suit you, otherwise do not hesitate to ask me.

joachimmetz commented:
@lprat thx, from a brief assessment the json file adds more questions than answers. I'm mainly looking for a simple explanation of the tagging rule, similar to a docstring. E.g.

```
Name: user_suspect_file
Description: tagging rule to flag potential suspicious files based on their extension as defined by MITRE ATTACK
...
URL: http://something
```

joachimmetz commented:
@lprat have a look at https://github.com/log2timeline/plaso/pull/2874/files#diff-24e74273f21e8afeef14136aee3a4e0b just a small description for every new tagging rule will be sufficient for now. I'll have a look to backfill the existing tagging rules.

joachimmetz commented:
@lprat thx for adding some descriptions, let me know when you are ready.

I've started adding tests and doing some clean up of the tag_windows.txt file https://github.com/log2timeline/plaso/pull/2883 

I had a look at https://mitre-attack.github.io/attack-navigator/enterprise/#layerURL=https%3A%2F%2Fraw.githubusercontent.com%2Flprat%2Fplaso%2Fmaster%2Fdata%2Fplaso_tagging_windows_coverage_mitre.json

It yields the following warning:
```
WARNING: Uploaded layer version (2.1) does not match Navigator's layer version (2.2). The layer configuration may not be fully restored.
```

I opt we move this file to `config/mitre-attack` equiv in combination with the README. It is becoming a bit more clear what you are implementing now. You want to map the Windows event tagging labels to the Mitre Attack ontology?

lprat commented:
Sorry for the delays between each response!
I fixed warning on version 2.2 (Mitre Attack).
Yes, I would like to map tagging labels to the Mitre Attack ontologoy.
Good idea to make test, because when i added description, i found errors in some rules...
Thanks

joachimmetz commented:
So https://github.com/log2timeline/plaso/pull/2883 was pushed, please rebase with HEAD. If you need help with that let me know.

lprat commented:
Hi,
Just to make sure you understand what you want. I get new tag file: tag_windows.txt pushed in #2883 and add my last modifications?

joachimmetz commented:
There is a warning for this PR now "This branch has conflicts that must be resolved"
since changes to Plaso codebase were made in #2883 that affect this PR,
these conflicts need to be resolved. One way of resolving them is to rebase or follow the github instructions if you're not too familiar with git.

If you allow edit by maintainers, I can make these changes as well if you're not comfortable with git.

lprat commented:
Ok, thank you for your explanations. I never did 'rebase' with git. Could you do , please? "Allow edits by maintainers" option is enable. Sorry, I will try to inquire about 'git rebase' for the next time. 
If you can't, I'll be watching this next week.

joachimmetz commented:
I've update the PR, have a look if everything is still fine. Was a bit of a merging challenge

joachimmetz commented:
Moved the MITRE ATT&CK configuration file and README into config directory since they are not data files directly used by plaso tools

joachimmetz commented:
Also please read "How should I reference the name ATT&CK?" on https://attack.mitre.org/resources/faq/

joachimmetz commented:
> Do you think that it's a problem in this context?

Per https://attack.mitre.org/resources/faq/, yes, we likely need to add a registered trademark symbol (®) as well. Unless there is another way or referring to this framework in a non-trademark way?

lprat commented:
Hi,
Sorry for my late response, again! 
I also saw reference "MITRE ATT&CK(TM)".
Do you want me to add at the beginning of the file tag_windows.txt: "#some tags may refer to the MITRE ATT&CK(TM)" ?

I think you deleted two files (plaso_tagging_windows_coverage_mitre.json, tag_windows.README) instead of moving them. But finally, for "tag_windows.README" it doesn't matter because "tag_windows.txt" already contains the same informations in the comments "#". For the json, I don't know.

joachimmetz commented:
I see I did not `git add` them after a file system move, should be fixed now.

joachimmetz commented:
Per their documentation each file seems to have one reference to MITRE ATT&CK. Now if (R), not (TM) that is a different thing, needs to be added, I"m not sure according to https://www.trademarknow.com/blog/tm-versus-r-whats-the-difference-and-why-does-it-matter:

```
It must only be used in the case of registered trademarks and by the owner or licensee. 
```

But unclear if we should consider ourselves to be a "licensee". Per https://attack.mitre.org/resources/faq/,

```
Your first references in writing must include "MITRE" preceding "ATT&CK®" - but subsequently should just reference "ATT&CK" (no registered trademark symbol required).
```

So let's include (R) for the first mention.

joachimmetz commented:
It is also quite ironic that it looks like Mitre is violating its own terms with "@MITREattack"

```
Do not modify the trademark, such as through hyphenation or abbreviation. For example, "ATT&CK'd!", "Plan-of-ATT&CK", "ATTK".
```

joachimmetz commented:
OK more inconsistency from Mitre https://github.com/mitre-attack/attack-navigator/issues/165

joachimmetz commented on data/tag_windows.txt:
```diff
@@ -1,74 +1,256 @@
+user_suspect_file
+  (data_type is 'fs:stat' OR data_type is 'fs:mactime:line') AND filename contains '/Users/' AND (filename contains '.exe' OR filename contains '.zip' OR filename contains '.rar' OR filename contains '.dll' OR filename contains '.cmd' OR filename contains '.rtf' OR filename contains '.pdf' OR filename contains '.doc' OR filename contains '.ppt' OR filename contains '.xls' OR filename contains '.v ...
```
`filename contains '/Users/'` this is going to break now with pyfsntfs being the default NTFS back-end.

lprat replied:
Can you explain me, the change in structure, filname field will not exist?

joachimmetz replied:
sry, good point forgot to relay that detail ;) 

the NTFS dfVFS back-end uses backslash (\), to be as close to the original as possible, instead of forward slash (/) which is used by TSK for everything. I'll need to re-assess how to do this in a less fragile way, but that is something a later point in time.

lprat replied:
Ok, it may impact other rules then present in tag_windows.txt ...

joachimmetz commented:
@lprat I've stared to do some clean up. Let's add the URL to the different techniques instead of mentioning the name of the framework so instead or T1195 add https://attack.mitre.org/techniques/T1195/

joachimmetz commented on data/tag_windows.txt:
@lprat can you clarify what you mean with this.

joachimmetz commented on data/tag_windows.txt:
@lprat was this what you meant?

joachimmetz commented on data/tag_windows.txt:
@lprat can you clarify what you mean with this.

joachimmetz commented on config/attack-navigator/plaso_tagging_windows_coverage_mitre.json:
```diff
@@ -0,0 +1,2445 @@
+{
```
IMHO it would be better if we could generate this file from tag_windows.txt to make sure it is kept up to date.

lprat replied:
Ok you're right, you can delete if you want, the important things is in tag_windows.txt.

joachimmetz commented:
> Mitre asking for clarification and if they can drop the administrative overhead.

Reply from Mitre that they are unwilling to drop the administrative burden on others so they can "enforce" their intellectual property and trademark status.

joachimmetz commented:
This makes me wonder if it easier to remove the attack-navigator config and just have the URLs in tag_windows.txt and make no reference to the name of this framework. To avoid having to deal with the administrative burden of ensuring the name is used correctly.

joachimmetz commented:
@lprat let's try to get this PR ready for merge. I've remove the config due previous conversation about the trademark.

lprat commented:
Descriptions of tagging rules was cleaned in "d769fe6c4c4e6be21dbfef0caaad0a9a8cd6ed4a"

joachimmetz commented:
> Descriptions of tagging rules was cleaned in "d769fe6c4c4e6be21dbfef0caaad0a9a8cd6ed4a"

Yeah there is more to be done. For sake of getting these changes merged and not making it a very long process of back and forths. I'll break this up in separate PRs, for the tagging rules that are clear to me. We can discuss the remainder after that. How does this sound?

lprat commented:
Ok, perfect.

lprat commented on data/tag_windows.txt:
Difficult to know the legitimacy. Even darkcomet can be legitimate because at the base it is not a malicious tool.
The question is whether a RAT tool is installed on the workstation for:
  - potentially find connection logs
  - ask the user for the legitimacy of the tool

joachimmetz replied:
Ack, then "Trojan" is unlikely to be the accurate terminology here. IMHO that indicates mal-intent.
Maybe remote_access_software / remote_access_application ?

lprat replied:
Ok, I let you make the choice!

lprat commented on data/tag_windows.txt:
I think you meant between: logon_suspect and login_attempt (on event 4624) ?
The difference is logontype:
-0 -> System Only
-4 -> batch
-5 -> service
-9 -> New credentials

lprat commented on data/tag_windows.txt:
This is the same problem as: https: //github.com/log2timeline/plaso/pull/2849/files/94e6e7ab318b55f0b8277fb2891d7f9cfc15fd17#r426216723? 
The fix is : "\Program Files\" ?

joachimmetz replied:
That is likely the same, I'll need to have a closer look at supporting paths. The current approach will fail on an implementation that generates a filename using `\`.

One options to add some helpers to indicate these are paths.
```
filename contains PATH_SEGMENTS(['Program Files'])
filename contains WINDOWSPATH('\Program Files\')
```

I'll need to look into this separately first before these rules can be integrated

lprat replied:
Ok, as soon as the functions are available WINDOWSPATH('') tell me so that I can implement.

joachimmetz replied:
Sure I will. There are some other priorities on my to list now, that need to be addressed first, but after that I"ll get back at porting these rules into Plaso (and corresponding changes).

joachimmetz commented on data/tag_windows.txt:
```diff
@@ -103,16 +329,294 @@ autorun
+  data_type is 'windows:evtx:record' AND (source_name is 'Microsoft-Windows-Iphlpsvc' OR source_name is 'Microsoft-Windows-Operating-System') AND event_identifier is 4200
+
+rdp_history
+  data_type is 'windows:registry:key_value' AND (key_path contains 'Software\\Microsoft\\Terminal Server Client\\Default' OR key_path contains 'Software\\Microsoft\\Terminal Server Client\\Servers')
```
There are separate winreg plugins for these: https://github.com/log2timeline/plaso/blob/master/plaso/parsers/winreg_plugins/terminal_server.py#L59 

What is the use case to label these?

lprat replied:
Yes, It doesnt need to keep this tag

joachimmetz commented on data/tag_windows.txt:
```diff
@@ -103,16 +329,294 @@ autorun
+rdp_history
+  data_type is 'windows:registry:key_value' AND (key_path contains 'Software\\Microsoft\\Terminal Server Client\\Default' OR key_path contains 'Software\\Microsoft\\Terminal Server Client\\Servers')
+
+smb_trace
```
Provide some context what you are labeling here? 

```
# Microsoft-Windows-SMBClient 30803 - Failed to establish a network connection
...
```

joachimmetz commented on data/tag_windows.txt:
```diff
@@ -103,16 +329,294 @@ autorun
+registry_modified
+  data_type is 'windows:evtx:record' AND source_name is 'Microsoft-Windows-Security-Auditing' AND event_identifier is 4657
+
+ip_change
```
Provide some context what you're labeling here

````
Microsoft-Windows-Iphlpsvc 4200 - Unable to read or write the NTE Context list for network adapter %2. IP interfaces on this network adapter may not be initialized completely.
```

joachimmetz replied:
@lprat this does not address my question. Please add some explanation what type of events this tagging rule is labeling and how to interpret the corresponding event log entries as "ip_change"

joachimmetz replied:
@lprat please correct this, this does not correspond with https://community.spiceworks.com/topic/775081-ms-server-event-id-4200-source-iphlpsvc-share-connectivity-help

lprat replied:
Sorry, I had not seen your questions in december. 
Tag "IP_change" is error! I had tagged with this name because I thought identified the event that indicated the IP address that the system takes when it connects to the network.

joachimmetz commented on data/tag_windows.txt:
```diff
@@ -1,12 +1,21 @@
   data_type is 'fs:stat' AND filename contains PATH('Windows/Tasks/At')
-  data_type is 'windows:evt:record' AND source_name is 'Security' AND event_identifier is 592
-  data_type is 'windows:evtx:record' AND source_name is 'Microsoft-Windows-Security-Auditing' AND event_identifier is 4688
+  (data_type is 'windows:evt:record' OR data_type is 'windows:evtx:record') AND source_name is 'Security' AND event_identifier is 592
```
Why this change? Isn't this ID only used by Windows versions that use Evt?

https://www.ultimatewindowssecurity.com/securitylog/encyclopedia/event.aspx?eventid=592

lprat replied:
Hi, sorry for event 592, you are right, it doesnt need to:
```OR data_type is 'windows:evtx:record')```

For event "4688", I put on one line with other event:
```
data_type is 'windows:evtx:record' AND source_name is 'Microsoft-Windows-Security-Auditing' AND (event_identifier is 4673 OR event_identifier is 4688)
```

joachimmetz commented:
@lprat can you address the comments and TODO a lot of the changes are not clear why they were made or are lacking context.

I did some clean up of your changes and moved some into a separate PR https://github.com/log2timeline/plaso/pull/3549

joachimmetz commented:
@lprat what is the status of this? Should I close this PR for now and reopen when you have the time work on it?

lprat commented:
Hi,
I'm sorry I forgot to consolidate the informations.
I propose to make a new PR for each tag change (i push soon PR to tag application_execution).
Thanks.
